Correct handling on modal dialog first hidden by Show(false) and restored by Show(true)

KaiZin2018 opened the pull request:
In the application pwsafe we do timely disabling of the dialogs shown by the application, independent if the main window or a modal dialog is shown. Recursively all Windows are hidden by calling Show(false). For modal dialog, running with own event loop in ShowModul(), the dialog is not closed in call of wxDialog::Show(false) [dialog_osx.cpp], but mark in m_modality is set back. When calling Show(true) for all Windows of the application, later time when having entered the password, the value of m_modality must be restored. When Show(true) is called again, the value must be restored, so wxDialogBase::EndDialog() is calling EndModal() instead of Hide() to avoid event loop is never finished.

KaiZin2018 committed:
Correct handling on modal dialog is first hidden by call Show(false) and later come back with Show(true). In that case the event loop will never be left in call of wxDialogBase::EndDialog().

vadz commented:
Wouldn't it be simpler (and hence better) to just avoid resetting `m_modality` to `NONE` in the first place for app modal dialogs? I.e. something like
```diff
diff --git a/src/osx/dialog_osx.cpp b/src/osx/dialog_osx.cpp
index aeef9e0f1a..cbe406a63d 100644
--- a/src/osx/dialog_osx.cpp
+++ b/src/osx/dialog_osx.cpp
@@ -133,18 +133,23 @@ bool wxDialog::Show(bool show)
 
     if ( !show )
     {
-        const int modalityOrig = m_modality;
-
-        // complete the 'hiding' before we send the event
-        m_modality = wxDIALOG_MODALITY_NONE;
-
-        switch ( modalityOrig )
+        switch ( m_modality )
         {
             case wxDIALOG_MODALITY_WINDOW_MODAL:
+                // complete the 'hiding' before we send the event
+                m_modality = wxDIALOG_MODALITY_NONE;
+
                 EndWindowModal(); // OS X implementation method for cleanup
                 SendWindowModalDialogEvent ( wxEVT_WINDOW_MODAL_DIALOG_CLOSED  );
                 break;
-            default:
+
+            case wxDIALOG_MODALITY_APP_MODAL:
+                // Keep m_modality unchanged in case Show(true) is called again
+                // later.
+                break;
+
+            case wxDIALOG_MODALITY_NONE:
+                // Nothing to do, the dialog is not really shown modally.
                 break;
         }
     }
```
?

But beyond this, I am not sure calling `Show(false)` for temporarily hiding modal dialogs is really supported. Does it really work for you under MSW and GTK?

KaiZin2018 commented:
> Wouldn't it be simpler (and hence better) to just avoid resetting `m_modality` to `NONE` in the first place for app modal dialogs? I.e. something like
> 
> ```diff
> diff --git a/src/osx/dialog_osx.cpp b/src/osx/dialog_osx.cpp
> index aeef9e0f1a..cbe406a63d 100644
> --- a/src/osx/dialog_osx.cpp
> +++ b/src/osx/dialog_osx.cpp
> @@ -133,18 +133,23 @@ bool wxDialog::Show(bool show)
>  
>      if ( !show )
>      {
> -        const int modalityOrig = m_modality;
> -
> -        // complete the 'hiding' before we send the event
> -        m_modality = wxDIALOG_MODALITY_NONE;
> -
> -        switch ( modalityOrig )
> +        switch ( m_modality )
>          {
>              case wxDIALOG_MODALITY_WINDOW_MODAL:
> +                // complete the 'hiding' before we send the event
> +                m_modality = wxDIALOG_MODALITY_NONE;
> +
>                  EndWindowModal(); // OS X implementation method for cleanup
>                  SendWindowModalDialogEvent ( wxEVT_WINDOW_MODAL_DIALOG_CLOSED  );
>                  break;
> -            default:
> +
> +            case wxDIALOG_MODALITY_APP_MODAL:
> +                // Keep m_modality unchanged in case Show(true) is called again
> +                // later.
> +                break;
> +
> +            case wxDIALOG_MODALITY_NONE:
> +                // Nothing to do, the dialog is not really shown modally.
>                  break;
>          }
>      }
> ```
> 
> ?
> 
> But beyond this, I am not sure calling `Show(false)` for temporarily hiding modal dialogs is really supported. Does it really work for you under MSW and GTK?

Your proposal isn't working (on macOS) - I tried as first guess. The file is dealing with macOS (OS X) only. I do not have the chance to check with MSW or GTK.

vadz commented:
I don't really understand why this doesn't work, it looks like it ought to. Do you understand/can you debug why it doesn't?

The trouble with making ad hoc fixes like this under a single platform is that nobody knows how exactly is this supposed to work and there are no tests checking it, so there is a high chance of breaking it when this code is modified the next time. We really need either a unit test reproducing what your application does and checking that it results in the expected behaviour or at least an test for doing this in the "dialogs" sample, so that it could be tested interactively.

Could you please try to make one? TIA!

KaiZin2018 committed:
Add test for pull request #2485 in menu "Dialogs" "Generic Dialogs" "Test Hide Modal Dialog".
Press "Hide Modal Dialog...", the dialog will disapear and a message box opens that has to be confirmewd to continue. With pressing "Cancel" the problem of not reactivated menu is visible. Pressing "Ok" will continue without problem as EndModal() is called here explicit.

KaiZin2018 commented:
> I don't really understand why this doesn't work, it looks like it ought to. Do you understand/can you debug why it doesn't?
> 
> The trouble with making ad hoc fixes like this under a single platform is that nobody knows how exactly is this supposed to work and there are no tests checking it, so there is a high chance of breaking it when this code is modified the next time. We really need either a unit test reproducing what your application does and checking that it results in the expected behaviour or at least an test for doing this in the "dialogs" sample, so that it could be tested interactively.
> 
> Could you please try to make one? TIA!

I have updated sample dialogs about a test of Hide modal dialog (in Dialogs/Generic Dialogs). In menu "Dialogs" / "Generic Dialogs" select "Test Hide Modal Dialog". Press "Hide Modal Dialog...", the dialog will disapear and a message box opens that has to be confirmewd to continue. With pressing "Cancel" the problem of not reactivated menu is visible. Pressing "Ok" will continue without problem as EndModal() is called here explicit.

vadz commented:
Thanks for the test, as I suspected, this doesn't work neither under MSW nor under GTK: under MSW the original dialog doesn't come back at all, under GTK it does flash briefly but immediately disappears.

Considering that currently this doesn't work under none of the major platforms, I think that it's safe to say that using modal dialogs like this is not supported. It's questionable whether it should be supported, IMO even if it is possible to support it under the other platforms, it would add complexity to the implementation for not much gain and it's not clear if it can be made to work everywhere. So I don't believe we want to merge this PR, sorry.

Concerning your original use case, I don't really understand why do you need to hide all the windows. Typically you would just disable them and, of course, simply showing a modal dialog already does this. But if you really want to hide the modal dialogs, you will need to dismiss them, i.e. call `EndModal()` rather than `Show(false)`.
